Add user management section (#14)
bufbuild/docs.buf.build: docs/bsr/user-management.md

---
id: user-management
title: User Management
---

# User Management

## Organization Roles

### Base resource roles

> *Base resource roles coming soon.*

> Every organization has a set of base resource roles that apply to all members of the organizatio. The following are the default roles:

Reviewer: ```suggestion
Every organization has a set of base resource roles that apply to all members of the organization. The following are the default roles:
```

- Repository: `write`
- Template: `write`
- Plugin: `write`

> Organization owners can modify the base resource roles depending on the requirements of the organization.

Reviewer: We should put this below the other roles, right? It's a more confusing concept, so maybe lead with the normal roles.

- Can view resources and users within the organization.

### Base resource roles

> *Base resource roles coming soon.*

Every organization has a set of base resource roles that apply to all members of the organization. The following are the default roles:

> | Repository | Template | Plugin |

Author: I liked the table view better than the bullet points

<img width="585" alt="CleanShot 2021-11-04 at 14 54 11@2x" src="https://user-images.githubusercontent.com/6278244/140402229-d0dea29d-71de-4201-9f44-965b3a8fa698.png">

Reviewer: Maybe capitalize these so they map to the resource role labels below?

Author: Maybe we shouldn't use backticks.. seems unecessary.

![CleanShot 2021-11-04 at 14 56 28@2x](https://user-images.githubusercontent.com/6278244/140402797-ba6646c4-8e1f-414d-8e74-daff55648db5.png)

Reviewer: Yeah 👍🏻

Author: What do you think, does it help with readability?

### **`Member`**

- Can view resources and users within the organization.

## Resource Roles

> *Resource roles coming soon.*

> Resource roles are explicit roles granted to a user on an individual resource, such as a repository, template or plugin.

Reviewer: Link to repository/template/plugin concepts here?

The most common use-cases for explicit resource roles are:

- Outside collaborators. This is useful when users outside your organization require access to specific resource(s) within the organization, but you do not want them to be a member of the organization.
- Elevated permissions for organization members. This is useful when the organization base resource roles are set to `read` and specific user(s) require `write` or `admin` permissions. 

> When computing the role on a resource, the highest role will take precedence. Example, an organization has `write` as the base repository role, but the user was mistakingly granted the `read` role on a specific repository. The final computed user role on the repository will be `write`.

Reviewer: ```suggestion
When computing the role on a resource, the highest role will take precedence. For example, an organization has `write` as the base repository role, but the user was mistakingly granted the `read` role on a specific repository. The final computed user role on the repository will be `write`.
```

Reviewer: We could make this sort of thing impossible by checking the base resource role of the organization at invitation time. Github does, we probably should too? Maybe a better example is that an organization admin can be invited to a repository explicitly, but their admin capacity gives them admin on the repository regardless.

Author: Good call. I originally had this as an explicit `admin` role on a repository to "override" the base repository `write` role. Will change it back to that example...

### **`Owner`**

- Unrestricted access to the resource.

### **`Admin`**

> - Can update the resource settings, visibility and deprecation notices.

Reviewer: Visibility updates aren't supported right now, we shouldn't include that here

---
id: user-management
title: User Management
---

# User Management

## Organization Roles

Every user that is part of an organization will have an explicit role. Note that users are unable to modify their own role. If you need to lower your access please have another organization user perform this action, or, leave the organization and request to be re-added with the desired role.
	
### Owner

- Users that require unrestricted access to the organization, its settings and all resources owned by the organization. 
- Can delete organization. All resources such as repositories, templates and plugins must be deleted before the organization can be deleted.

### Admin

- Can manage user roles, except owners.
- Can add and delete resources such as [repositories](../bsr/overview.md#module), [templates](../bsr/remote-generation/concepts/#template) and [plugins](../bsr/remote-generation/concepts/#plugin).

### Member

> - Can view resources and users within the organization.

Reviewer 2: Technically, members have write access to resources (as you say below).

It might be better to say something along the lines of

```suggestion
- Can view the organization and its members
- Have the [Base resource role](<LINK TO HEADING HERE>) over the organizations resources, which defaults to [Write](<LINK TO RESOURCE WRITE PERMISSION HERE>).
```

And then in Base resource roles, you can say exactly what you do say.

### Base resource roles

> > *Base resource roles coming soon.*

Reviewer 2: I would say that *configurable* base resource roles are coming soon.

Every organization has a set of base resource roles that apply to all members of the organization. The following are the default roles:

| Repository | Template | Plugin |
|:--|:--|:--|
| **Write**  | **Write** | **Write** |

Organization owners can modify the base resource roles depending on the requirements of the organization.

## Resource Roles

> *Resource roles coming soon.*

Resource roles are explicit roles granted to a user on an individual resource, such as a repository, template or plugin.

The most common use-cases for explicit resource roles are:

- Outside collaborators. This is useful when users outside your organization require access to specific resource(s) within the organization, but you do not want them to be a member of the organization.
- Elevated permissions for organization members. This is useful when the organization base resource roles are set to **Read** and specific user(s) require the **Write** or **Admin** role for specific resource(s). 

> When computing the role on a resource, the highest role will take precedence. For example, an organization has **Write** as the base repository role, and the user was granted the **Admin** role on a specific repository. The final computed user role on the repository will be **Admin**.

Reviewer 2: Do we want to disambiguate resource roles — which you can be granted by organization membership — from "granting direct access" or "adding outside collaborators"? 

It might be useful to define the ways you can interact with a resource — which is true already for user-owners and any organization members — separate from defining how you'll be able to configure resource-to-user specific roles in the future.

Author: Not sure if this is what you mean, but I added a blurb to this section to expand on the user- and organization-owned resources and the default roles users have .. which is a segue into why we'd give additional permissions to a user over a resource.

Let me know if this is what you were going for?

https://deploy-preview-14--buf.netlify.app/bsr/user-management/#resource-roles

### Owner

- Unrestricted access to the resource.

### Admin

- Can update the resource settings and deprecation notices.
- Can manage resource roles, except owners.

> - Can delete the resource.

Reviewer 2: Can admins delete the resource? This used to be only owners; just confirming that we want that behavior to change. I think I originally thought all destructive behavior (name changing, deletion) would be limited to owners.

Author: You're right, deletion should be limited to the owner resource role.

### Write

> - Can perform modifying operations, except for deletes.

Reviewer 2: Might want to be more specific here — this just means like `push`ing and making branches and tags, right? Not updating the resource name, which might also be considered a modifying operation.

Again, I'm not actually sure what's enforced by the new authz, but I'm of the opinion that they should only be able to write to the content of the resource, rather than the resource as a DB entity ¯\_(ツ)_/¯

Author: > Might want to be more specific here

Good point, updated with a few concrete examples.
